File: apps/api-nest/src/shared/ids.ts
import { v7 as uuidv7 } from 'uuid';

/**
 * Generate a new UUIDv7 for use as a primary key.
 *
 * UUIDv7 is time-ordered (RFC 9562), which preserves B-tree index locality
 * compared to random UUIDv4 values. This mirrors the Go implementation's
 * `internal/shared/ids.New()`.
 *
 * IDs are generated in the application layer rather than by PostgreSQL so
 * that a caller can know the ID of a row before inserting it.
 */
export function newId(): string {
  return uuidv7();
}

File: apps/api-legacy/internal/shared/ids/ids.go
// Package ids centralizes ID generation for all domain modules.
//
// All primary keys are UUIDv7 (RFC 9562): time-ordered, which improves
// B-tree index locality compared to random UUIDv4 values.
package ids

import "github.com/google/uuid"

// New returns a new UUIDv7 as a string, for use as a TEXT primary key.
func New() string {
	id, err := uuid.NewV7()
	if err != nil {
		// uuid.NewV7 only fails if the system source of randomness fails,
		// which is unrecoverable. Fall back to the package-level generator
		// (which panics on the same condition) rather than returning "".
		return uuid.Must(uuid.NewRandom()).String()
	}
	return id.String()
}
